Refactor `./pants` and fix bugs in `ci.py` in preparation for Python 3 support

Eric-Arellano opened the pull request:
This makes several improvements and fixes that will result in a smaller diff for https://github.com/pantsbuild/setup/pull/32.

Fixes:
- Fix missing `write_config(updated_config)` when setting up `pants_version`, meaning when running `./ci.py --unspecified` we wouldn't actually remove the `pants_version` from `pants.ini`. This hotfixes https://github.com/pantsbuild/setup/pull/35.
- Properly restore `pants.ini` in CI, even if the subprocesses failed. Due to https://github.com/pantsbuild/setup/pull/34 now properly propagating return codes, the context manager would exit prematurely before restoring the original `pants.ini`.

Refactors:
- Check the Python interpreter is on the PATH and exit with error message if not.
- Extract Python interpreter setup into a dedicated function `determine_pants_runtime_python_version()` for a smaller diff with #32.
- Improve `pants.ini` parsing:
   - Extract `get_pants_ini_config_value()` to generify reading `pants.ini`.
   - Extend `pants.ini` parsing to recognize both `:` and `=`, along with optional spaces before and/or after the delimiter. These were all valid `ini` entries, but our script only handled one specific case.
   - Respect `#` in front of `pants_version` by requiring the line to start exactly with `pants_version`.
- Refactor `pants.ini` parsing to use variables for readability.
- Add `green` and `die` functions with colored output
   - Use `die` if interpreter cannot be resolved
   - Use `green` upon successful creation of a new virtual environment the first time

Eric-Arellano committed:
Refactor for a smaller diff

Eric-Arellano committed:
Extract out determine_pants_runtime_python_version()

Make it into a function for smaller diff + check it's a valid interpreter.

Eric-Arellano committed:
Extract out get_pants_ini_config_value()

Eric-Arellano committed:
Properly cleanup pants.ini if process fails

Eric-Arellano committed:
Don't use tab character

Bad copy and paste

Eric-Arellano committed:
Make sed even more readable!

cosmicexplorer commented on pants:
```diff
@@ -45,12 +43,30 @@ function tempdir {
+  optional_space="[[:space:]]*"
+  prefix="^${config_key}${optional_space}${valid_delimiters}${optional_space}"
+  remove_prefix=" s#${prefix}##p"
+  sed -ne "/${prefix}/${remove_prefix}" pants.ini
```
The way sed commands work in this case is that `/<pattern>/ s#<replace-pattern>#<replacement>#<flags>` says "for lines matching `<pattern>`, run the following `s` command -- replace `<replace-pattern>` in the line with `<replacement>`, printing or doing whatever according to `<flags>`". In this case I think the sed command itself shouldn't be split up, because the actual `s` command is what is doing things here and it makes it feel spooky otherwise. Something like:
```bash
  sed -ne "/${prefix}/ s#${prefix}##p" pants.ini
```
I think makes what's happening very clear ("when encountering prefix, remove it"). The use of `#` is something I like to do -- people tend to use `/` but that tends to clash a lot with things I want to replace, imo. The sed info pages are very nice but we hopefully shouldn't be doing anything complex enough to require them in a script we distribute like this.

Eric-Arellano replied:
Agreed, this is much better. Thanks for the thorough explanation! Have not used sed much.

cosmicexplorer commented on pants:
```diff
@@ -45,12 +43,30 @@ function tempdir {
+  else
+    echo "${python_bin_name} could not be found on your PATH! Pants will not work." 1>&2
+    exit 1
+  fi
```
This is readable enough, but could also be:
```bash
  if ! which "${python_bin_name}"; then
    echo "${python_bin_name} could not be found on your PATH! Pants will not work." 1>&2
    exit 1
  fi
```
It will produce an extra `<x> not found` line in the error case, but I think that's ok / what we want here.

Eric-Arellano replied:
Yes I like this more. Thanks.

cosmicexplorer replied:
With the above, I was actually thinking you could remove the `>/dev/null` and also remove the `echo` line below the `if`.

cosmicexplorer commented on pants:
```diff
@@ -65,40 +81,39 @@ function bootstrap_venv {
 cd "$(cd "$(dirname ${BASH_SOURCE[0]})" && pwd -P)"
-pants_dir=$(bootstrap_pants)
+python="$(determine_pants_runtime_python_version)"
+pants_dir="$(bootstrap_pants ${python})"
```
```suggestion
pants_dir="$(bootstrap_pants "${python}")"
```
The inside of a `"$()"` substitution is a clean environment in terms of quoting, so the usual considerations with double-quoting dereferences apply. This isn't always necessary, but it could matter at some point, so it's useful to be consistent, I think.

Eric-Arellano replied:
Didn't realize you can put multiple double quotes within double quotes! Good to know.

Eric-Arellano committed:
Add Danny's suggestions

stuhood commented on pants.ini:
```diff
@@ -1,6 +1,6 @@
 [GLOBAL]
-pants_version:1.15.0.dev3
-plugins:[
+pants_version : 1.15.0.dev3
```
Is this a related change? I've generally seen `key: value` , but presumably the matching code works for any combination.

Eric-Arellano replied:
Yes, this unfortunately is the result of us using in `ci.py` `ConfigParser` to rewrite back our `pants.ini` after the tests. It doesn't allow us to specify the spacing beyond if we want no spaces or a space on both sides. 

Originally, our regex parsing wasn't flexible enough to allow a space before the `:` (even though it's perfectly valid `ini` syntax), so we had to use no spaces. Now that we can use either style, I changed back to the spaces because it's `ConfigParser`'s default and I think is easier to read.

stuhood commented on pants:
```diff
@@ -45,12 +43,28 @@ function tempdir {
+  config_key="$1"
+  valid_delimiters="[:=]"
+  optional_space="[[:space:]]*"
+  prefix="^${config_key}${optional_space}${valid_delimiters}${optional_space}"
```
It looks like the old regex used to allow a leading optional space before the key... which is odd? It would seem to make more sense the way you have it here.

cosmicexplorer commented on pants:
```diff
@@ -65,40 +79,39 @@ function bootstrap_venv {
-  )
-  if [[ -n "${pants_version}" ]]
-  then
+  pants_version=$(get_pants_ini_config_value 'pants_version')
```
```suggestion
  pants_version="$(get_pants_ini_config_value 'pants_version')"
```

Eric-Arellano committed:
Respond to review comments

Eric-Arellano committed:
Add back error message when python interpreter not found

Without it, we silently fall when the interpreter cannot be resolved.

Eric-Arellano committed:
Properly use Danny's suggestion of de-duping which() call

Eric-Arellano committed:
Introduce die function and colored output

cosmicexplorer reviewed (approved):
Thanks so much for iterating on this!

Eric-Arellano committed:
Define green, not blue

We don't introduce all colors from common.sh to try to keep things simpler. We're more likely to use green (success) than blue, so leave off blue and add green.

stuhood commented on pants:
```diff
@@ -33,24 +31,58 @@ SETUPTOOLS_REQUIREMENT="setuptools==5.4.1"
-function tempdir {
-  mktemp -d "$1"/pants.XXXXXX
+function determine_pants_runtime_python_version {
+  python_bin_name="${PYTHON:-python2.7}"
```
Since `PYTHON` remains a "parameter" to this entire script, having it near the top of the file like it was before seems like good documentation of that fact.

Eric-Arellano committed:
Add green output for successful venv and add back global Python variable at top to make it explicit we take an env var optionally

cosmicexplorer reviewed (approved):
Approving again for the `PYTHON_BIN_NAME` changes! Thanks!
